### PR TITLE
Create helper methods to ensure stream reads are read correctly to the expected length

### PR DESCRIPTION
**Pull request by peppy**
Based loosely around the solution proposed [here](https://stackoverflow.com/a/221941).

The only thing I'm not 100% on is whether we should use the `EndOfStreamException` here as there could be a super edge case where a stream reads past the end of the expected length? Not sure that will ever happen though.

- [x] Depends on #5022

### CODE REVIEW

**Comment by peppy** [comment]
Blocking temporarily to add test coverage.

File: osu.Framework/Platform/TcpIpcProvider.cs [hunk 1]
```diff
@@ -182,21 +182,21 @@ private async Task send(Stream stream, IpcMessage message)
...
-            int readBytes = await stream.ReadAsync(header.AsMemory(), cancellationToken).ConfigureAwait(false);
+            const int header_length = sizeof(int);
 
-            if (readBytes < header.Length)
-                throw new EndOfStreamException();
+            byte[] header = new byte[header_length];
 
-            int len = BitConverter.ToInt32(header, 0);
-            if (len == 0)
+            int read = await stream.ReadAsync(header.AsMemory(), cancellationToken).ConfigureAwait(false);
```
Comment (bdach) [inline osu.Framework/Platform/TcpIpcProvider.cs:189]: Shouldn't this be a

```csharp
byte[] header = await stream.ReadBytesToArrayAsync(header_length, cancellationToken).ConfigureAwait(false);
```

call instead?

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        /// Read the full content of a stream.
+        /// </summary>
+        /// <param name="stream">The stream to read.</param>
+        /// <returns>The full byte content.</returns>
+        public static byte[] ReadAllBytesToArray(this Stream stream)
+        {
+            if (stream.CanSeek)
+            {
+                stream.Seek(0, SeekOrigin.Begin);
+                Debug.Assert(stream.Length < int.MaxValue);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:24]: Due to the int cast in the next line. Can remove if you don't think it's required.

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                return stream.ReadBytesToArray((int)stream.Length);
+            }
+
+            return stream.ReadAllRemainingBytesToArray();
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:28]: Can do, was a toss up for me.

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+        /// <summary>
+        /// Read specified length from current position in stream.
+        /// </summary>
+        /// <param name="stream">The stream to read.</param>
+        /// <param name="length">The length to read.</param>
+        /// <returns>The full byte content.</returns>
+        public static byte[] ReadBytesToArray(this Stream stream, int length)
+        {
+            byte[] buffer = new byte[16 * 1024];
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:57]: Matches .net internals

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Framework/Extensions/StreamExtensions.cs:24]: Oh, I see. I think I'd rather change `ReadBytesToArray(Async)` to accept length as `long` rather than `int`. Seems like it will generate less friction against the underlying stream API anyhow.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:24]: The base API only allows `int` as far as I could see :(.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Framework/Extensions/StreamExtensions.cs:24]: Oh because `Read()` takes `int` and `Position`/`Length` takes `long`... Kinda icky.

You could arguably handle that mismatch internally - since `ReadBytesToArray()` will only ever read at most 16 * 1024 bytes at a time, you could change that method to accept `long` and safely cast down the result of the `Math.Min()` operation to `int`. But given that `int.MaxValue` bytes is about 2GB of data then maybe no real need to worry about it because hopefully nobody is reading that much...?

**Comment by bdach** [comment]
For kicks I've also written a quick and terrible `MemoryStream`-like implementation that only ever reads one byte at a time, to see if it would pass the tests added here - and the tests do indeed pass. Not sure if it's worth including but I'll leave here for posterity at least?

<details>
<summary>the slowest stream implementation to ever exist</summary>

```csharp
        private class SingleByteReadingMemoryStream : Stream
        {
            private readonly byte[] buffer;

            public SingleByteReadingMemoryStream(byte[] buffer)
            {
                this.buffer = buffer;
            }

            #region Reading

            public override bool CanRead => true;

            public override int Read(byte[] destination, int offset, int count)
            {
                if (count < 1 || position >= Length)
                    return 0;

                destination[offset] = buffer[position];
                position += 1;
                return 1;
            }

            #endregion

            #region Writing

            public override bool CanWrite => false;

            public override void Write(byte[] destination, int offset, int count) => throw new NotSupportedException();

            public override void Flush() => throw new NotSupportedException();

            public override void SetLength(long value) => throw new NotSupportedException();

            #endregion

            #region Seek support

            public override long Length => buffer.Length;

            private long position;

            public override long Position
            {
                get => position;
                set => Seek(value, SeekOrigin.Begin);
            }

            public override bool CanSeek => true;

            public override long Seek(long offset, SeekOrigin origin)
            {
                switch (origin)
                {
                    case SeekOrigin.Begin:
                        position = offset;
                        break;

                    case SeekOrigin.Current:
                        position += offset;
                        break;

                    case SeekOrigin.End:
                        position = Length + offset;
                        break;
                }

                return position;
            }

            #endregion
        }
```

</details>

The async test cases take 11 seconds (!!!) to complete with that implementation. I don't even want to think about the TPL thrashing that is causing.

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                while ((read = stream.Read(buffer, 0, Math.Min(remainingRead, buffer.Length))) > 0)
+                {
+                    ms.Write(buffer, 0, read);
+                    remainingRead -= read;
+                }
+
+                if (remainingRead != 0)
+                    throw new EndOfStreamException();
+
+                return ms.GetBuffer();
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:74]: that's why there's a length check, yep.

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:24]: yeah, i doubt it'll be an issue. we can handle if someone hits the exception.

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:74]: to clarify, this usage in fact doesn't require the specific length check as the initial capacity was specified (which sets the internal buffer size).

**Review by bdach — DISMISSED** [review]

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            byte[] buffer = new byte[buffer_size];
+
+            using (var ms = new MemoryStream())
+            {
+                int read;
+
+                while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
+                    ms.Write(buffer, 0, read);
+
+                return ms.ToArray();
```
Comment (smoogipoo) [inline osu.Framework/Extensions/StreamExtensions.cs:133]: Any reason for this to use `.ToArray()` when the others use `.GetBuffer()`?

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            byte[] buffer = new byte[buffer_size];
+
+            using (var ms = new MemoryStream())
+            {
+                int read;
+
+                while ((read = await stream.ReadAsync(buffer.AsMemory(), cancellationToken).ConfigureAwait(false)) > 0)
+                    await ms.WriteAsync(buffer, 0, read, cancellationToken).ConfigureAwait(false);
+
+                return ms.ToArray();
```
Comment (smoogipoo) [inline osu.Framework/Extensions/StreamExtensions.cs:154]: Here too.

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        /// <returns>The full byte content.</returns>
+        public static async Task<byte[]> ReadAllRemainingBytesToArrayAsync(this Stream stream, CancellationToken cancellationToken = default)
+        {
+            byte[] buffer = new byte[buffer_size];
+
+            using (var ms = new MemoryStream())
+            {
+                int read;
+
+                while ((read = await stream.ReadAsync(buffer.AsMemory(), cancellationToken).ConfigureAwait(false)) > 0)
```
Comment (smoogipoo) [inline osu.Framework/Extensions/StreamExtensions.cs:151]: Any reason for using `.AsMemory()`?

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+        /// <summary>
+        /// Read the full content of a seekable stream.
+        /// </summary>
+        /// <remarks>
+        /// For a non-seekable stream, use <see cref="ReadAllRemainingBytesToArray"/> instead.
+        /// </remarks>
+        /// <param name="stream">The stream to read.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>The full byte content.</returns>
```
Comment (smoogipoo) [inline osu.Framework/Extensions/StreamExtensions.cs:42]: I'd prefer that these methods had `<exception>` docs for their special cases, which show up in Rider:
image
As opposed to a non-explicit tidbit in the remarks.

I don't mean just this method too, but `ReadBytesToArray()` as well because, as far as I know, `Stream.Read(length)` defines a _maximum_ read length, whereas the implementations below throw `EndOfStreamException`s in the same scenario.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 8 above)
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:151]: Already touched on [in this discussion](https://github.com/ppy/osu-framework/pull/5024#discussion_r805152756).

The internal buffer is not equal to the actual read length as it uses the standard .NET buffer expansion (doubles in size each time it needs to). Requesting `GetBuffer` would return an array with zero bytes at the end in most cases, which we don't want here.

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 6 above)
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:133]: Already touched on [in this discussion](https://github.com/ppy/osu-framework/pull/5024#discussion_r805152756).

The internal buffer is not equal to the actual read length as it uses the standard .NET buffer expansion (doubles in size each time it needs to). Requesting `GetBuffer` would return an array with zero bytes at the end in most cases, which we don't want here.

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 8 above)
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:151]: No need, was a remnant of a previous version.

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 9 above)
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:43]: Have added.

File: osu.Framework/Extensions/StreamExtensions.cs (hunk 7 above)
Comment (peppy) [reply osu.Framework/Extensions/StreamExtensions.cs:158]: Similar case to the one explained above. The only special case where I'm using `GetBuffer` is where the `new MemoryStream(length)` ctor is used, along with a length check (and throw) before returning the buffer.

File: osu.Framework/Extensions/StreamExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        public static async Task<byte[]> ReadAllRemainingBytesToArrayAsync(this Stream stream, CancellationToken cancellationToken = default)
+        {
+            byte[] buffer = new byte[buffer_size];
+
+            using (var ms = new MemoryStream())
+            {
+                int read;
+
+                while ((read = await stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false)) > 0)
+                    await ms.WriteAsync(buffer, 0, read, cancellationToken).ConfigureAwait(false);
```
Comment (smoogipoo) [inline osu.Framework/Extensions/StreamExtensions.cs:156]: Is there a reason for not using `stream.CopyToAsync()`? Seems like it does everything here plus uses the global array pool.

Would prefer to not reinvent the wheel.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
I have my gripes with this but let's see how it goes with the current implementation... What I was going to post is as follows, but I don't want to start a discussion about it:

I really don't know about this... The SO thread is based on the presumption that you can't trust `stream.Length`, but this code is explicitly saying you _can_ trust it (at least in the case of `ReadAllBytesToArray()`, which event asserts as much).  
Whether you can trust it in the case of e.g. network streams, is an answer I can't provide, because I simply don't know every edge case that exists. While it's true that [`Stream.Length`](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.length?view=net-6.0) throws an exception if not seekable which is covered by the guard in `ReadAllBytesToArray()`, could data be streamed in and `.Length` not represent the true length of the stream? Sounds possible to me, or at least there's preventing this.

Personal preference would be for the EOS exception to not exist and the methods to read as much as they can until EOS.

**Comment by peppy** [comment]
> Personal preference would be for the EOS exception to not exist and the methods to read as much as they can until EOS.

Main reason for not doing this is it results in a second copy of the array being required, unless you want to branch for that exception case and add more complexity (something I was trying to avoid here because I agree with not wanting to reinvent the wheel, the point I wish these methods existed somewhere else and we didn't need to write them).

**Comment by smoogipoo** [comment]
Dunno. Take that as you may. I don't think copying is a huge deal - at the point you're reading so much data that that's a point of concern, you probably shouldn't be doing this in the first place. But you could also return sliced `Memory` instances instead if that's that big of a concern.

**Review by smoogipoo — APPROVED** [review]

**Comment by peppy** [comment]
This is being used for loading of all resources (aka `DllResourceStore`), which can be hundreds of megabytes for a game. So I thought it was a point of concern already. I was consciously benchmarking memory usage when writing these methods.
